experimental/sample_web_static/index.html

```diff
@@ -0,0 +1,225 @@
+  </canvas>
+
+  <br>
+  <div style="border:2px solid #000000; background-color: #CCCCCC; padding: 8px; color: #111111" width="400px" height="300px">
+    <button id="predictButton" disabled onclick="predictDigit()">Predict handwritten digit</button>
+    <br>
+    Prediction result: <div id="predictionResult"></div>
+  </div>
+
+  <script>
+    let setupSample;
+    let cleanupSample;
+    let runSample;
+    let state;
+    let predictionResultElement;
+    let initialized = false;
+
+    var Module = {
+      print: function(text) {
+        console.log(text);
+      },
+      printErr: function(text) {
+        console.error(text);
+      },
+      onRuntimeInitialized: function() {
+        console.log("WebAssembly module onRuntimeInitialized()");
+
+        setupSample = Module.cwrap("setup_sample", "number", []);
+        cleanupSample = Module.cwrap("cleanup_sample", null, ["number"]);
+        runSample = Module.cwrap("run_sample", "number", ["number", "number"]);
+
+        state = setupSample();
+
+        predictionResultElement = document.getElementById("predictionResult");
+        document.getElementById("predictButton").disabled = false;
+
+        initialized = true;
+      },
+      // https://emscripten.org/docs/api_reference/module.html#Module.noInitialRun
+      noInitialRun: true,
+    };
+
+    function predictDigit() {
+      const rawImageData = getRescaledCanvasData();
+      const imageDataBuffer = preprocessImageData(rawImageData);
+
+      result = runSample(state, imageDataBuffer);
+      if (result != -1) {
+        predictionResultElement.innerHTML = result;
+      } else {
+        predictionResultElement.innerHTML = "Error";
+      }
+
+      Module._free(imageDataBuffer);
+    }
+
+    // https://becominghuman.ai/passing-and-returning-webassembly-array-parameters-a0f572c65d97
+    // https://developers.google.com/web/updates/2018/03/emscripting-a-c-library#get_an_image_from_javascript_into_wasm
+    function preprocessImageData(rawImageData) {
+      // * getImageData() returns a Uint8ClampedArray with RGBA image data
+      // * this MNIST model takes tensor<1x28x28x1xf32> with grayscale pixels
+      //   in [0.0, 1.0]
+
+      // This conversion is terrible, but this is a toy demo with a small image
+      // Hopefully there aren't any logic / iteration order issues...
+      const imagePixelCount = 28 * 28;
+      const typedArray = new Float32Array(imagePixelCount);
+      for (let y = 0; y < 28; ++y) {
+        for (let x = 0; x < 28; ++x) {
+          const typedIndex = y * 28 + x;
+          const rawIndex = 4 * (y * 28 + x) + 3;  // Assume colorSpace srgb
+          typedArray[typedIndex] = rawImageData.data[rawIndex] / 255.0;
+        }
+      }
+
+      // Copy into WASM heap.
+      // Note: we could have done the conversion in-place, but this is demo code
+      const buffer = Module._malloc(imagePixelCount * Float32Array.BYTES_PER_ELEMENT);
```

Comment:
you should be able to alloc this once and then write into each time and pass the same pointer - sure it's demo code, but no reason to be so bad :)

Reply:
Ah... yep. Done :)